xds/src/main/java/com/linecorp/armeria/xds/client/endpoint/ClusterManager.java

```diff
@@ -110,33 +128,30 @@ public void snapshotUpdated(ListenerSnapshot listenerSnapshot) {
             return;
         }
         final RouteSnapshot routeSnapshot = listenerSnapshot.routeSnapshot();
-        if (routeSnapshot == null) {
-            return;
-        }
-
-        // it is important that the entries are added in order of ClusterSnapshot#index
-        // so that the first matching route is selected in #selectNow
-        final ImmutableMap.Builder<String, ClusterEntry> mappingBuilder =
-                ImmutableMap.builder();
-        final Map<String, ClusterEntry> oldEndpointGroups = clusterEntries;
-        for (ClusterSnapshot clusterSnapshot : routeSnapshot.clusterSnapshots()) {
+        final List<ClusterSnapshot> clusterSnapshots =
+                routeSnapshot != null ? routeSnapshot.clusterSnapshots() : ImmutableList.of();
+        final Map<String, ClusterEntry> oldClusterEntries = clusterEntries;
+        // ImmutableMap is used because it is important that the entries are added in order of
+        // ClusterSnapshot#index so that the first matching route is selected in #selectNow
+        final ImmutableMap.Builder<String, ClusterEntry> mappingBuilder = ImmutableMap.builder();
+        for (ClusterSnapshot clusterSnapshot : clusterSnapshots) {
             if (clusterSnapshot.endpointSnapshot() == null) {
                 continue;
             }
             final String clusterName = clusterSnapshot.xdsResource().name();
-            ClusterEntry clusterEntry = oldEndpointGroups.get(clusterName);
+            ClusterEntry clusterEntry = oldClusterEntries.get(clusterName);
             if (clusterEntry == null) {
-                clusterEntry = new ClusterEntry(clusterSnapshot, this, eventLoop);
-            } else {
-                clusterEntry.updateClusterSnapshot(clusterSnapshot);
+                clusterEntry = new ClusterEntry(eventLoop, localCluster);
             }
+            clusterEntry.updateClusterSnapshot(clusterSnapshot);
+            clusterEntry.addListener(ignored -> notifyListeners(), true);
             mappingBuilder.put(clusterName, clusterEntry);
         }
         final ImmutableMap<String, ClusterEntry> newClusterEntries = mappingBuilder.build();
         clusterEntries = newClusterEntries;
         this.listenerSnapshot = listenerSnapshot;
```

Comment:
In practice 1) `listenerSnapshot` is only used for equality checks when signalling 2) the `listenerSnapshot` will eventual be consistent with endpoints from `clusterEntries`, so I didn't think it was important that the consistency between the two is important.

Having said this, I agree it can be confusing for other maintainers so I've added a wrapper class at https://github.com/line/armeria/pull/5808/commits/6d87f23c97d1de472a8c8170c2c66ba266b84805